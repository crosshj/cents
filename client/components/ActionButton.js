import React from 'react';
import { Menu, ChildButton, MainButton } from 'react-mfb';

const menuProps = {
    effect: 'slidein',
    position: 'br',
    method: 'click'
};


function ActionButton({ onChoose }) {
    const childMenuClick = (which) => {
        onChoose && onChoose(which)
        const mainIcon = document && document.querySelector('.mfb-component__main-icon--active');
        /* istanbul ignore next */
        mainIcon && mainIcon.click();
        // ^^^ hard to test this due to how react-mfb is written
    };

    return (
        <div className='floating-action-button'>
            <Menu {...menuProps}>
                <MainButton
                    iconResting="a-icon fa fa-plus"
                    iconActive="a-icon fa fa-times"
                />
                <ChildButton
                    onClick={() => childMenuClick('dollar')}
                    icon="a-icon fa fa-dollar"
                    label="Add a new expense"
                />
                <ChildButton
                    onClick={() => childMenuClick('group')}
                    icon="a-icon fa fa-align-justify"
                    label="Create a group"
                />
                <ChildButton
                    onClick={() => childMenuClick('refresh')}
                    icon="a-icon fa fa-refresh"
                    label="Update application"
                />
            </Menu>
        </div>
    );
}

export default ActionButton;