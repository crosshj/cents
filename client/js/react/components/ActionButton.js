import React from 'react';
import { Menu, ChildButton, MainButton } from 'react-mfb';

const menuProps = {
    effect: 'slidein',
    position: 'br',
    method: 'click'
};

const childMenuClick = () => {
    console.log('Action Button child click: handler should be passed by parent!');
    document.querySelector('.mfb-component__main-icon--active').click();
};

function ActionButton() {
    return (
        <div className='floating-action-button'>
            <Menu {...menuProps}>
                <MainButton
                    iconResting="a-icon fa fa-plus"
                    iconActive="a-icon fa fa-times"
                />
                <ChildButton
                    onClick={childMenuClick}
                    icon="a-icon fa fa-dollar"
                    label="Add a new expense"
                />
                <ChildButton
                    onClick={childMenuClick}
                    icon="a-icon fa fa-align-justify"
                    label="Create a group"
                />
                <ChildButton
                    onClick={childMenuClick}
                    icon="a-icon fa fa-refresh"
                    label="Update application"
                />
            </Menu>
        </div>
    );
}

export default ActionButton;