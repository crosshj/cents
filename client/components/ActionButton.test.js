import React from 'react';
import ActionButton from './ActionButton';
import renderer from 'react-test-renderer';

describe('floating action button', () => {

  it('renders initial view', () => {
    const component = renderer.create(
      <ActionButton />,
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('changes menu state to open when main button clicked', () => {
    const component = renderer.create(
      <ActionButton />,
    );
    const mainMenu = component.root.findByProps({ className: 'mfb-slidein mfb-component--br'})
    var mainMenuState = mainMenu.props['data-mfb-state'];
    expect(mainMenuState).toBe('closed');

    const button = component.root.findByProps({className: 'mfb-component__button--main'});
    const eventMock = { preventDefault: jest.fn() };
    button.props.onClick(eventMock);

    mainMenuState = mainMenu.props['data-mfb-state'];
    expect(mainMenuState).toBe('open');
  });

  it('calls onChoose function when child button is clicked', () => {
    const onChoose = jest.fn();

    const component = renderer.create(
      <ActionButton onChoose={onChoose}/>,
    );

    const childButtons = component.root.findAllByProps({className: 'mfb-component__button--child'});
    childButtons.forEach(child => child.props.onClick());

    expect(onChoose).toBeCalledWith('dollar');
    expect(onChoose).toBeCalledWith('group');
    expect(onChoose).toBeCalledWith('refresh');

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});


