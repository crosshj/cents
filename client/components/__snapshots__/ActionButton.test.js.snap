// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`floating action button calls onChoose function when child button is clicked 1`] = `
<div
  className="floating-action-button"
>
  <ul
    className="mfb-slidein mfb-component--br"
    data-mfb-state="closed"
    data-mfb-toggle="click"
  >
    <li
      className="mfb-component__wrap"
    >
      <a
        className="mfb-component__button--main"
        href="#"
        onClick={[Function]}
      >
        <i
          className="mfb-component__main-icon--resting a-icon fa fa-plus"
        />
        <i
          className="mfb-component__main-icon--active a-icon fa fa-times"
        />
      </a>
      <ul
        className="mfb-component__list"
      >
        <li>
          <a
            className="mfb-component__button--child"
            data-mfb-label="Add a new expense"
            onClick={[Function]}
          >
            <i
              className="mfb-component__child-icon a-icon fa fa-dollar"
            />
          </a>
        </li>
        <li>
          <a
            className="mfb-component__button--child"
            data-mfb-label="Create a group"
            onClick={[Function]}
          >
            <i
              className="mfb-component__child-icon a-icon fa fa-align-justify"
            />
          </a>
        </li>
        <li>
          <a
            className="mfb-component__button--child"
            data-mfb-label="Update application"
            onClick={[Function]}
          >
            <i
              className="mfb-component__child-icon a-icon fa fa-refresh"
            />
          </a>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`floating action button renders initial view 1`] = `
<div
  className="floating-action-button"
>
  <ul
    className="mfb-slidein mfb-component--br"
    data-mfb-state="closed"
    data-mfb-toggle="click"
  >
    <li
      className="mfb-component__wrap"
    >
      <a
        className="mfb-component__button--main"
        href="#"
        onClick={[Function]}
      >
        <i
          className="mfb-component__main-icon--resting a-icon fa fa-plus"
        />
        <i
          className="mfb-component__main-icon--active a-icon fa fa-times"
        />
      </a>
      <ul
        className="mfb-component__list"
      >
        <li>
          <a
            className="mfb-component__button--child"
            data-mfb-label="Add a new expense"
            onClick={[Function]}
          >
            <i
              className="mfb-component__child-icon a-icon fa fa-dollar"
            />
          </a>
        </li>
        <li>
          <a
            className="mfb-component__button--child"
            data-mfb-label="Create a group"
            onClick={[Function]}
          >
            <i
              className="mfb-component__child-icon a-icon fa fa-align-justify"
            />
          </a>
        </li>
        <li>
          <a
            className="mfb-component__button--child"
            data-mfb-label="Update application"
            onClick={[Function]}
          >
            <i
              className="mfb-component__child-icon a-icon fa fa-refresh"
            />
          </a>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
